### ISSUE TITLE
HEAppE returns internal error status code (500) when using an expired session code

### ISSUE THREAD

**Issue by lupreCSC** [It4innovations/HEAppE#1]
When making a request with an expired session code, our deployment of HEAppE v4.2.0 internally encounters an exception
```
HEAppE.Exceptions.External.SessionCodeNotValidException: Expired
```
which propagates to the client as a internal server error (status code 500) response. This is at least the case for a request to `/heappe/JobManagement/CreateJob`.

It would be helpful for client error handling if a more meaningful status code like `401 Unauthorized` would be returned from the API server.

Extract from server logs
```
ERROR 2024-05-20 09:25:49 HEAppE.RestApi.ExceptionMiddleware - Session code '<code removed>' already expired at '5/20/2024 6:34:10 AM'. 
HEAppE.Exceptions.External.SessionCodeNotValidException: Expired
   at HEAppE.BusinessLogicTier.Logic.UserAndLimitationManagement.UserAndLimitationManagementLogic.GetUserForSessionCode(String sessionCode) in /src/BusinessLogicTier/Logic/UserAndLimitationManagement/UserAndLimitationManagementLogic.cs:line 77
   at HEAppE.ServiceTier.JobManagement.JobManagementService.CreateJob(JobSpecificationExt specification, String sessionCode) in /src/ServiceTier/JobManagement/JobManagementService.cs:line 42
   at HEAppE.RestApi.Controllers.JobManagementController.CreateJob(CreateJobByProjectModel model) in /src/RestApi/Controllers/JobManagementController.cs:line 54
   at lambda_method1698(Closure, Object, Object[])
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at HEAppE.RestApi.ExceptionMiddleware.InvokeAsync(HttpContext context) in /src/RestApi/ExceptionMiddleware.cs:line 72

```

As a side note, it also looks like the expiration date for the status code above is given in UTC while all other times are in local time (but that is not relevant for this issue).

**Comment by jkonvicka** [comment]
Dear @lupreCSC,
Thank you for your message and for bringing this issue to our attention.

Our dev team confirm that your concern is valid. When making a request with an expired session code in HEAppE v4.2.0, an internal exception, ``HEAppE.Exceptions.External.SessionCodeNotValidException: Expired``, is encountered, which propagates to the client as an internal server error (status code 500). The expected behavior is to return a ``401 Unauthorized`` status code.

Additionally, I have noted your observation regarding the inconsistent date-time format. Although this is not directly related to the issue at hand, it is a valid point, and we will address it accordingly.

I really appreciate your feedback and I'm pleased to inform you that a fix for these issues will be included in the next release, which we are currently working on.

Thank you for your understanding and patience.

Best regards,
Jakub Konvicka

**Comment by lupreCSC** [comment]
Hello @jkonvicka,

thanks, fixing this will be very helpful in the future. I noticed now in general that some of the status codes seem a bit off. For example, I earlier today has an error with an invalid job creation request (due to missing some required fields in the request), which had a status code of `404` (Not Found) when `400` (Bad Request) would probably have been more appropriate. In general there's a bit of a lack in the documentation of what possible errors responses the API might give, which would be helpful to have :) Are there any plans on improving this in the future?

**Comment by jkonvicka** [comment]
Hi @lupreCSC, 
I’m working on the documentation improvements right now. And yes, there will be documented REST API endpoints with status codes. Also I will look to the all status codes - to be sure that are appropriate. 

Thanks a lot!

**Comment by jkonvicka** [comment]
Hi @lupreCSC, 
the fix was included in the new release HEAppE (V4.3.0).
